### PR TITLE
BUG: Improve memory usage with openpyxl

### LINKED ISSUE #41681: BUG: df.to_excel() with openpyxl engine doesn't use write-optimized mode, resulting in higher memory consumption
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

---

#### Problem description

openpyxl has fairly high memory consumption by default. To address this for write-only operations, they recommend two things:

1. Install the lxml package, which is more optimized than the default xml parser. 
2. Use the write_only mode on the workbook.

In `pandas/pandas/io/excel/_openpyxl.py` we have the [a comment](https://github.com/pandas-dev/pandas/blob/d662e9740a8cd738970cce97e341640b00cadcf8/pandas/io/excel/_openpyxl.py#L64) which states intent - `# Create workbook object with default optimized_write=True.` 

However openpyxl depreciated the `optimized_write` option in [version 2.4](https://openpyxl.readthedocs.io/en/stable/changes.html?highlight=optimized_write%3DTrue.#a1-2016-04-11) and now recommends setting `write_only=True` on workbook creation. Pandas requires 2.6 so the result is that writes to excel use more memory than necessary. Switching to `write_only=True` has other ramifications, like only being able to write out once, so would need to be tested.

#### Proposed solution

1. Print out a warning about `lxml` being missing if writing to excel in an environment that doesn't have it installed. 
2. Use the `write_only` mode if possible.

### PR DESCRIPTION
**Pull request by jordanrmerrick**
- [x] closes #41681
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them

### CODE REVIEW

File: pandas/__init__.py [hunk 1]
```diff
@@ -1,5 +1,11 @@
 # flake8: noqa
 
+import locale
+locale.setlocale(locale.LC_ALL, 'C.UTF-8')
+
+import locale
+locale.setlocale(locale.LC_ALL, 'C.UTF-8')
```
Comment (jbrockmendel) [inline pandas/__init__.py:7]: why is this needed at all, much less done twice?

Comment (jordanrmerrick) [reply pandas/__init__.py]: This was my mistake, I'm not sure why this was added to __init__.  Thank you for pointing it out!

File: pandas/io/excel/_openpyxl.py [hunk 2]
```diff
@@ -61,17 +61,23 @@ def __init__(
             self.sheets = {name: self.book[name] for name in self.book.sheetnames}
 
         else:
-            # Create workbook object with default optimized_write=True.
-            self.book = Workbook()
+            try:
+                # Sheets are not automatically created in the workbook
+                self.book = Workbook(write_only=True)
+            except ImportError:
```
Comment (datapythonista) [inline pandas/io/excel/_openpyxl.py:67]: If you want to know if `lxml` is installed, use `import lxml`. This `ImportError` can be generated by different missing dependencies.

File: pandas/io/excel/_openpyxl.py [hunk 3]
```diff
@@ -61,17 +61,23 @@ def __init__(
...
+                print("Warning: lxml is not installed")
```
Comment (datapythonista) [inline pandas/io/excel/_openpyxl.py:68]: If we want to send warning to the user, we better use proper warnings, this is not the way.

File: pandas/io/excel/_openpyxl.py [hunk 4]
```diff
@@ -61,17 +61,23 @@ def __init__(
...
-                self.book.remove(self.book.worksheets[0])
+                if self.book.worksheets:
+                    self.book.remove(self.book.worksheets[0])
 
     def save(self):
         """
         Save workbook to disk.
         """
-        self.book.save(self.handles.handle)
+        # TODO: Handle errors from saving more than once
```
Comment (datapythonista) [inline pandas/io/excel/_openpyxl.py:79]: What are we supposed to do with this TODO? Are you fixing it later in this PR, or do we need to create an issue?

**Review by datapythonista — CHANGES_REQUESTED** [review]
Thanks for the contribution @jordanrmerrick 

Since this is fixing a bug, the first thing would be to write a test that it fails with the current implementation. So we can see if the test is fixed after making the code changes. In this case can be a bit trickier, since for what I understand the bug is too high memory consumption. But in any case, if we don't have the test, how do we know this implementation makes sense?

And we  also need a release note.

File: pandas/io/excel/_openpyxl.py (hunk 3 above)
Comment (jordanrmerrick) [reply pandas/io/excel/_openpyxl.py:68]: > If we want to send warning to the user, we better use proper warnings, this is not the way.

I thought so, can I ask if there is a standard that pandas uses for passing errors to the user? I think this also brings up an interesting point; if `write_only` mode is unavailable, should that throw an error and stop the code, or just initialize a read/write workbook instead?

File: pandas/io/excel/_openpyxl.py (hunk 4 above)
Comment (jordanrmerrick) [reply pandas/io/excel/_openpyxl.py:79]: That was a TODO for me! I'm committing it today.

Comment (jordanrmerrick) [reply pandas/io/excel/_openpyxl.py:79]: > What are we supposed to do with this TODO? Are you fixing it later in this PR, or do we need to create an issue?

That was a TODO for me! I'm committing it today.

File: pandas/io/excel/_openpyxl.py (hunk 3 above)
Comment (datapythonista) [reply pandas/io/excel/_openpyxl.py:68]: Python has a `warning` module, you can grep in the code for samples.

My understanding is that we are currently using unnecessary memory when reading Excel files. I think what we want is to try to save memory, but if it's not possible continue to open the file anyway. Warning the user that could be saving memory in the operation by installing a library seems reasonable. But I don't think it makes sense to fail if the user has enough memory for the operation.

Comment (jordanrmerrick) [reply pandas/io/excel/_openpyxl.py:68]: Your understanding is correct! I agree, a warning seems reasonable enough. I'm mainly asking because I'm new to contributing to pandas and still trying to get a feel for best practices within the library :)

File: pandas/io/excel/_openpyxl.py (hunk 2 above)
Comment (jordanrmerrick) [reply pandas/io/excel/_openpyxl.py:67]: > If you want to know if `lxml` is installed, use `import lxml`. This `ImportError` can be generated by different missing dependencies.

Sounds good, I'll make that change now.

File: pandas/io/excel/_openpyxl.py [hunk 5]
```diff
@@ -29,12 +29,17 @@ class OpenpyxlWriter(ExcelWriter):
     engine = "openpyxl"
     supported_extensions = (".xlsx", ".xlsm")
 
+    # Because this is used a few times in the class,
+    # it's declared as a variable
+    # Maybe move to within __init__?
+
```
Comment (jordanrmerrick) [inline pandas/io/excel/_openpyxl.py:35]: Forgot to remove this, I ended up moving the variable into `__init__`. Should it remain there?

File: pandas/io/excel/_openpyxl.py [hunk 6]
```diff
@@ -410,6 +438,17 @@ def _convert_to_protection(cls, protection_dict):
...
+    def _convert_to_excel_format(row, col):
+        col_name = ""
+
+        while col > 0:
+            mod = (col - 1) % 26
+            col_name = str(chr(65 + mod)) + col_name
+            col = (col - mod) // 26
+
+        return "%s%d" % (col_name, row)
+
```
Comment (jordanrmerrick) [inline pandas/io/excel/_openpyxl.py:451]: Due to the nature of how merging cells in `write_only` mode works, we need the row and column in Excel notation rather than its index. This converts a row and column number to excel format (e.g. A1, ZZ99, etc.).

File: pandas/io/excel/_openpyxl.py [hunk 7]
```diff
@@ -448,52 +487,123 @@ def write_cells(
...
+                        last_col = startcol + cell.mergeend + 1
+
+                        for row in range(first_row, last_row + 1):
+                            for col in range(first_col, last_col + 1):
+                                if row == first_row and col == first_col:
+                                    # Ignore first cell. It is already handled.
+                                    continue
+                                xcell = wks.cell(column=col, row=row)
+                                for k, v in style_kwargs.items():
+                                    setattr(xcell, k, v)
```
Comment (jordanrmerrick) [inline pandas/io/excel/_openpyxl.py:557]: Haven't finished this part yet. I'm still trying to figure out a way to apply the style to the other cells in the merged range. I'm not entirely sure there is one, but I'll spend more time prodding around.

If I can't find a reasonable workaround, does this have serious ramifications? I imagine probably not given that it's just styling the underlying cells, but maybe I'm missing something.

**Comment by jordanrmerrick** [comment]
The above commit won't work right now, I'm just adding it so the progress is visible. I still need to change how the cells are styled before it should work.

**Comment by rhshadrach** [comment]
@jordanrmerrick Friendly ping to see if you're interested in continuing this PR.

**Comment by mroeschke** [comment]
Thanks for the contribution, but it appears this PR has gone stale. Closing due to inactivity but if you're interested in continuing we would be happy to reopen.

**Comment by jordanrmerrick** [comment]
> Thanks for the contribution, but it appears this PR has gone stale. Closing due to inactivity but if you're interested in continuing we would be happy to reopen.

Hello, unfortunately I had to take an extended hiatus from pretty much all my personal coding work. I'm thankfully able to pick it up again, and this is a big priority of mine!

**Comment by jordanrmerrick** [comment]
Hi, posting some updates here: 

I'm still working on this, but progress has been slow. The way `write_only` workbooks work is causing complications with how to add styles (fonts, colors, etc.) to cells in a way that works with the existing methods in Pandas' code. Merged cells are also proving to be quite tricky. 

It's taking so long to add to this because I think I'm going to have to rewrite a lot of Pandas' implementation of openpyxl and I'm trying to plan it out beforehand. Sorry about that! I should be pushing ~some~ code in the next day or so.

**Comment by jreback** [comment]
closing as stale but ping if you want to continue
